### ISSUE TITLE
Bad serialization of response when using @EnableWebMvc

### ISSUE THREAD

**Issue by theHilikus** [zalando/problem-spring-web#334]
Whenever I introduce `@EnableWebMvc` on a configuration class, the response of a bad request is of type `application/problem+json`, as it should, but it has all the fields: stacktrace, parameters, etc etc. When I comment out the annotation, it serializes properly. is this a bug? it sounds to me like it's _not_, but is it something that can be improved in the library so that it works in both scenarios?

With `@EnableWebMvc`
```{
    "cause": null,
    "stackTrace": [
        {
            "methodName": "noHandlerFound",
            "fileName": "DispatcherServlet.java",
            "lineNumber": 1253,
            "className": "org.springframework.web.servlet.DispatcherServlet",
            "nativeMethod": false
        },
        {
            "methodName": "doDispatch",
            "fileName": "DispatcherServlet.java",
            "lineNumber": 1017,
            "className": "org.springframework.web.servlet.DispatcherServlet",
            "nativeMethod": false
        },
        {
            "methodName": "doService",
            "fileName": "DispatcherServlet.java",
            "lineNumber": 942,
            "className": "org.springframework.web.servlet.DispatcherServlet",
            "nativeMethod": false
        },
        {
            "methodName": "processRequest",
            "fileName": "FrameworkServlet.java",
            "lineNumber": 1005,
            "className": "org.springframework.web.servlet.FrameworkServlet",
            "nativeMethod": false
        },
        {
            "methodName": "doGet",
            "fileName": "FrameworkServlet.java",
            "lineNumber": 897,
            "className": "org.springframework.web.servlet.FrameworkServlet",
            "nativeMethod": false
        },
        {
            "methodName": "service",
            "fileName": "HttpServlet.java",
            "lineNumber": 645,
            "className": "javax.servlet.http.HttpServlet",
            "nativeMethod": false
        },
        {
            "methodName": "service",
            "fileName": "FrameworkServlet.java",
            "lineNumber": 882,
            "className": "org.springframework.web.servlet.FrameworkServlet",
            "nativeMethod": false
        },
        {
            "methodName": "service",
            "fileName": "HttpServlet.java",
            "lineNumber": 750,
            "className": "javax.servlet.http.HttpServlet",
            "nativeMethod": false
        },
        {
            "methodName": "handle",
            "fileName": "ServletHolder.java",
            "lineNumber": 873,
            "className": "org.eclipse.jetty.servlet.ServletHolder",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "ServletHandler.java",
            "lineNumber": 1623,
            "className": "org.eclipse.jetty.servlet.ServletHandler$CachedChain",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "WebSocketUpgradeFilter.java",
            "lineNumber": 214,
            "className": "org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "ServletHandler.java",
            "lineNumber": 1610,
            "className": "org.eclipse.jetty.servlet.ServletHandler$CachedChain",
            "nativeMethod": false
        },
        {
            "methodName": "doFilterInternal",
            "fileName": "RequestContextFilter.java",
            "lineNumber": 99,
            "className": "org.springframework.web.filter.RequestContextFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "OncePerRequestFilter.java",
            "lineNumber": 107,
            "className": "org.springframework.web.filter.OncePerRequestFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "ServletHandler.java",
            "lineNumber": 1610,
            "className": "org.eclipse.jetty.servlet.ServletHandler$CachedChain",
            "nativeMethod": false
        },
        {
            "methodName": "doFilterInternal",
            "fileName": "FormContentFilter.java",
            "lineNumber": 92,
            "className": "org.springframework.web.filter.FormContentFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "OncePerRequestFilter.java",
            "lineNumber": 107,
            "className": "org.springframework.web.filter.OncePerRequestFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "ServletHandler.java",
            "lineNumber": 1610,
            "className": "org.eclipse.jetty.servlet.ServletHandler$CachedChain",
            "nativeMethod": false
        },
        {
            "methodName": "doFilterInternal",
            "fileName": "HiddenHttpMethodFilter.java",
            "lineNumber": 93,
            "className": "org.springframework.web.filter.HiddenHttpMethodFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "OncePerRequestFilter.java",
            "lineNumber": 107,
            "className": "org.springframework.web.filter.OncePerRequestFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "ServletHandler.java",
            "lineNumber": 1610,
            "className": "org.eclipse.jetty.servlet.ServletHandler$CachedChain",
            "nativeMethod": false
        },
        {
            "methodName": "doFilterInternal",
            "fileName": "CharacterEncodingFilter.java",
            "lineNumber": 200,
            "className": "org.springframework.web.filter.CharacterEncodingFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "OncePerRequestFilter.java",
            "lineNumber": 107,
            "className": "org.springframework.web.filter.OncePerRequestFilter",
            "nativeMethod": false
        },
        {
            "methodName": "doFilter",
            "fileName": "ServletHandler.java",
            "lineNumber": 1610,
            "className": "org.eclipse.jetty.servlet.ServletHandler$CachedChain",
            "nativeMethod": false
        },
        {
            "methodName": "doHandle",
            "fileName": "ServletHandler.java",
            "lineNumber": 540,
            "className": "org.eclipse.jetty.servlet.ServletHandler",
            "nativeMethod": false
        },
        {
            "methodName": "handle",
            "fileName": "ScopedHandler.java",
            "lineNumber": 146,
            "className": "org.eclipse.jetty.server.handler.ScopedHandler",
            "nativeMethod": false
        },
        {
            "methodName": "handle",
            "fileName": "SecurityHandler.java",
            "lineNumber": 548,
            "className": "org.eclipse.jetty.security.SecurityHandler",
            "nativeMethod": false
        },
        {
            "methodName": "handle",
            "fileName": "HandlerWrapper.java",
            "lineNumber": 132,
            "className": "org.eclipse.jetty.server.handler.HandlerWrapper",
            "nativeMethod": false
        },
        {
            "methodName": "nextHandle",
            "fileName": "ScopedHandler.java",
            "lineNumber": 257,
            "className": "org.eclipse.jetty.server.handler.ScopedHandler",
            "nativeMethod": false
        },
        {
            "methodName": "doHandle",
            "fileName": "SessionHandler.java",
            "lineNumber": 1700,
            "className": "org.eclipse.jetty.server.session.SessionHandler",
            "nativeMethod": false
        },
        {
            "methodName": "nextHandle",
            "fileName": "ScopedHandler.java",
            "lineNumber": 255,
            "className": "org.eclipse.jetty.server.handler.ScopedHandler",
            "nativeMethod": false
        },
        {
            "methodName": "doHandle",
            "fileName": "ContextHandler.java",
            "lineNumber": 1345,
            "className": "org.eclipse.jetty.server.handler.ContextHandler",
            "nativeMethod": false
        },
        {
            "methodName": "nextScope",
            "fileName": "ScopedHandler.java",
            "lineNumber": 203,
            "className": "org.eclipse.jetty.server.handler.ScopedHandler",
            "nativeMethod": false
        },
        {
            "methodName": "doScope",
            "fileName": "ServletHandler.java",
            "lineNumber": 480,
            "className": "org.eclipse.jetty.servlet.ServletHandler",
            "nativeMethod": false
        },
        {
            "methodName": "doScope",
            "fileName": "SessionHandler.java",
            "lineNumber": 1667,
            "className": "org.eclipse.jetty.server.session.SessionHandler",
            "nativeMethod": false
        },
        {
            "methodName": "nextScope",
            "fileName": "ScopedHandler.java",
            "lineNumber": 201,
            "className": "org.eclipse.jetty.server.handler.ScopedHandler",
            "nativeMethod": false
        },
        {
            "methodName": "doScope",
            "fileName": "ContextHandler.java",
            "lineNumber": 1247,
            "className": "org.eclipse.jetty.server.handler.ContextHandler",
            "nativeMethod": false
        },
        {
            "methodName": "handle",
            "fileName": "ScopedHandler.java",
            "lineNumber": 144,
            "className": "org.eclipse.jetty.server.handler.ScopedHandler",
            "nativeMethod": false
        },
        {
            "methodName": "handle",
            "fileName": "HandlerWrapper.java",
            "lineNumber": 132,
            "className": "org.eclipse.jetty.server.handler.HandlerWrapper",
            "nativeMethod": false
        },
        {
            "methodName": "handle",
            "fileName": "Server.java",
            "lineNumber": 505,
            "className": "org.eclipse.jetty.server.Server",
            "nativeMethod": false
        },
        {
            "methodName": "handle",
            "fileName": "HttpChannel.java",
            "lineNumber": 370,
            "className": "org.eclipse.jetty.server.HttpChannel",
            "nativeMethod": false
        },
        {
            "methodName": "onFillable",
            "fileName": "HttpConnection.java",
            "lineNumber": 267,
            "className": "org.eclipse.jetty.server.HttpConnection",
            "nativeMethod": false
        },
        {
            "methodName": "succeeded",
            "fileName": "AbstractConnection.java",
            "lineNumber": 305,
            "className": "org.eclipse.jetty.io.AbstractConnection$ReadCallback",
            "nativeMethod": false
        },
        {
            "methodName": "fillable",
            "fileName": "FillInterest.java",
            "lineNumber": 103,
            "className": "org.eclipse.jetty.io.FillInterest",
            "nativeMethod": false
        },
        {
            "methodName": "run",
            "fileName": "ChannelEndPoint.java",
            "lineNumber": 117,
            "className": "org.eclipse.jetty.io.ChannelEndPoint$2",
            "nativeMethod": false
        },
        {
            "methodName": "runTask",
            "fileName": "EatWhatYouKill.java",
            "lineNumber": 333,
            "className": "org.eclipse.jetty.util.thread.strategy.EatWhatYouKill",
            "nativeMethod": false
        },
        {
            "methodName": "doProduce",
            "fileName": "EatWhatYouKill.java",
            "lineNumber": 310,
            "className": "org.eclipse.jetty.util.thread.strategy.EatWhatYouKill",
            "nativeMethod": false
        },
        {
            "methodName": "tryProduce",
            "fileName": "EatWhatYouKill.java",
            "lineNumber": 168,
            "className": "org.eclipse.jetty.util.thread.strategy.EatWhatYouKill",
            "nativeMethod": false
        },
        {
            "methodName": "run",
            "fileName": "EatWhatYouKill.java",
            "lineNumber": 126,
            "className": "org.eclipse.jetty.util.thread.strategy.EatWhatYouKill",
            "nativeMethod": false
        },
        {
            "methodName": "run",
            "fileName": "ReservedThreadExecutor.java",
            "lineNumber": 366,
            "className": "org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread",
            "nativeMethod": false
        },
        {
            "methodName": "runJob",
            "fileName": "QueuedThreadPool.java",
            "lineNumber": 698,
            "className": "org.eclipse.jetty.util.thread.QueuedThreadPool",
            "nativeMethod": false
        },
        {
            "methodName": "run",
            "fileName": "QueuedThreadPool.java",
            "lineNumber": 804,
            "className": "org.eclipse.jetty.util.thread.QueuedThreadPool$Runner",
            "nativeMethod": false
        },
        {
            "methodName": "run",
            "fileName": "Thread.java",
            "lineNumber": 748,
            "className": "java.lang.Thread",
            "nativeMethod": false
        }
    ],
    "type": "about:blank",
    "title": "Not Found",
    "status": "NOT_FOUND",
    "detail": "No handler found for GET /genericd",
    "instance": null,
    "parameters": {},
    "message": "Not Found: No handler found for GET /genericd",
    "localizedMessage": "Not Found: No handler found for GET /genericd",
    "suppressed": []
}
```
Without `@EnableWebMvc`
```
{
    "title": "Not Found",
    "status": 404,
    "detail": "No handler found for GET /genericd"
}
```

**Comment by whiskeysierra** [comment]
I guess you have some `WebMvcConfigurer` that overrides the standard `MappingJackson2HttpMessageConverter`.

**Comment by theHilikus** [comment]
not explicitly. i don't know what the annotation does internally. this is all I have

```
@Configuration
@EnableWebMvc
public class MyConfig implements WebMvcConfigurer {
    @Override
    public void addResourceHandlers(ResourceHandlerRegistry registry) {
        System.out.println("Adding resource handlers");
    }
}
```

and note that it's not the class itself. If I leave that exact same class and just comment out `@EnableWebMvc`, serialization works fine again.

**Comment by whiskeysierra** [comment]
https://stackoverflow.com/questions/45734108/spring-boot-not-using-configured-jackson-objectmapper-with-enablewebmvc

**Comment by whiskeysierra** [comment]
https://docs.spring.io/spring-boot/docs/current/reference/html/howto-spring-mvc.html#howto-switch-off-default-mvc-configuration

**Comment by theHilikus** [comment]
I knew that it disables autoconfiguration for webMVC. What i meant in my last message is that I don't know for sure whether it "overrides the standard MappingJackson2HttpMessageConverter" like you said or if it's something else that's disabling the normal registration of jackson modules or if the modules are registered but not used.
Regardless, do you know any proper/clean way to use the library while using `@EnableWebMvc` or is the library just not compatible with this configuration? I tried to play over the weekend with registering the modules in this override of `WebMvcConfigurer`:
```
public void configureMessageConverters(List<HttpMessageConverter<?>> converters)
```
and it works but I couldn't find a way to make such a bean conditional on the `@EnableWebMvc` annotation. I figured there must be a way to do it the regular way, as in the documentation (with @Bean) when enablewebmvc is not present and a way to do it with `configureMessageConverters` when it is present, but `@ConditionalOnMissingBean(WebMvcConfigurationSupport.class)` didn't work. It makes sense to me to have something in the documentation about the "official/recommended" way to register zalando's lib when using `@EnableWebMvc`, hence why I created the ticket. Just like you have an official instruction when using springboot on top of spring

**Comment by whiskeysierra** [comment]
What exactly do you need `@EnableWebMvc` for?

**Comment by theHilikus** [comment]
in my particular case it is to reduce memory footprint since the library might run in an embedded device and autowiring pulls a lot of classes/feature that are not needed; but in a general case, it seems like a valid case to want to use `@EnableWebMvc` with `problem-spring-web`, no?

**Comment by whiskeysierra** [comment]
Sure, it's a valid use case. But you would need to configure the message converter and the Jackson object mapper by hand, since the default is not being used.

**Comment by theHilikus** [comment]
whether you use `@EnableWebMvc` or not, you have to configure the jackson object mapper by hand to use problem-spring-web. i.e. there's no autoconfiguration as part of the library: if you have a normal springboot project, you just need to create the two beans for the jackson modules (ProblemModule and ConstraintViolationProblemModule). if you disable autoconfiguration there's another way (the one discussed here). 
So my point is, just like you document that you need to create those two beans to use the lib, wouldn't it make sense to document how to use the lib in the case that webmvc autoconfiguration is disabled? or you think it's not a common-enough usecase? if you don't think it's worth it, then at least documenting that webmvc autoconfiguration is a requirement would be nice

**Comment by whiskeysierra** [comment]
> if you don't think it's worth it, then at least documenting that webmvc autoconfiguration is a requirement would be nice

Agreed. Would you like to propose a pull request?

**Comment by theHilikus** [comment]
yes. 
And on that same note, would you be open to a PR for autoconfiguring the library so that you just need to add the lib to the classpath and then it works? I don't know if I will be able to do it since I've never created spring autoconfigurations but I might give it a try if you think it would be accepted

**Comment by whiskeysierra** [comment]
If it's a separate artifact/module that users can pick and choose, then sure, why not?

**Comment by whiskeysierra** [comment]
I released your change as [0.25.0](https://github.com/zalando/problem-spring-web/releases/tag/0.25.0) to central just now. Thanks for your contribution :tada: and sorry that it took so long.

**Comment by IncPlusPlus** [comment]
Leaving this here as a note for people who might run into this issue. The improper serialization of responses will likely also occur in tests that use the `@WebMvcTest` annotation. The solution that worked for me was to add `@Import(ProblemJacksonAutoConfiguration.class)` to my test classes.
